### PR TITLE
Add helper function for deleting a session

### LINKED ISSUE #8: Delete session if OAuth2 access token is no longer valid
I'm wondering how to best deal with the following scenario:

- User visits site, is not logged in and goes through the OAuth2 login flow.
- User now has a session containing an OAuth2 access token.
- Time passes.
- User visits site again. The middleware passes the user through on account of the session still being valid, but when the site uses the access token from the session to make an OAuth2-protected request it founds out the access token is no longer valid.

At this point I think the session is effectively dead, and the session cookie should be destroyed so the user can go through the authentication flow again. There's a couple of ways this might happen:

1. The application could destroy the cookie itself (this is what I do currently). Downside: it needs to know the exact cookie name and path to do so, which is tied to the implementation of `wai-middleware-auth`.
1. The middleware could expose a helper function for generating the header required to delete the cookie. That keeps the knowledge for dealing with the cookie within this package.
1. The application sends a 401 error response, and the middleware picks up on this and adds a header on the response deleting the session cookie.

I particularly like the last option because it keeps all session management concerns on the layer of the middleware. On the other hand the behavior might be too magical. Maybe it could be an option?

I'd love your thoughts. If there's an interest I might follow up with a PR.

### PR DESCRIPTION
**Pull request by jwoudenberg**
Resolves #8.

### CODE REVIEW

File: src/Network/Wai/Middleware/Auth.hs [hunk 1]
```diff
@@ -297,3 +299,11 @@ getAuthUser = Vault.lookup userKey . vault
...
 waiMiddlewareAuthVersion = Paths.version
 
+-- | Get a response header to delete the users current session.
+--
+-- @since 0.2.0
+getDeleteSessionHeader :: AuthSettings -> IO Header
+getDeleteSessionHeader AuthSettings {..} = do
+  secretKey <- asGetKey
+  -- Delete the cookie by setting an expiry date far in the past.
+  saveCookieValue secretKey asStateKey (-365 * 24 * 3600) ("" :: S.ByteString)
```
Comment (lehins) [inline src/Network/Wai/Middleware/Auth.hs]: Setting age in the past looks hacky, plus with `saveCookieValue` the value, even if empty, is being encrypted, so in the end cookie will contains some value that decodes into empty bytestring, which is not what we want.

I'd rather use a helper function like this:
```haskell

deleteCookieValue
  :: Binary value
  => IO Header
deleteCookieValue name = do
  return
    ( "Set-Cookie"
    , toByteString $
      renderSetCookie
        def
          { setCookieName = name
          , setCookieValue = ""
          , setCookiePath = Just "/"
          , setCookieHttpOnly = True
          , setCookieExpires = Just $ UTCTime (fromGregorian 1970 01 01) 0
          })
```

Comment (jwoudenberg) [reply src/Network/Wai/Middleware/Auth.hs]: That seems much nicer. Pushed a change to implement this change!

Comment (lehins) [reply src/Network/Wai/Middleware/Auth.hs]: I am sorry, it was pretty late for me last night too. Could you get rid of the `IO` in `deleteCookieValue` and in `getDeleteSessionHeader`, those are all pure computations. Other than that looks great and almost ready to be merged

Comment (jwoudenberg) [reply src/Network/Wai/Middleware/Auth.hs]: Oops, completely missed that too! Fixed it.

Where are you based?

Comment (lehins) [reply src/Network/Wai/Middleware/Auth.hs]: Minsk, Belarus.

Comment (jwoudenberg) [reply src/Network/Wai/Middleware/Auth.hs]: Cool! Working from Cambridge, UK myself. Nice to meet you!

Comment (lehins) [reply src/Network/Wai/Middleware/Auth.hs]: It is my pleasure :)
I was in your neck of the woods just a couple of days ago, at Haskell Exchange in London.

Comment (jwoudenberg) [reply src/Network/Wai/Middleware/Auth.hs]: Yeah, I just saw the videos for that they posted. Looks like there were many great talks. I'm considering going next year.

**Pull request merged**